File: c/BWT.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void swap(char *x, char*y)
{
  char t = *x; 
  *x = *y;
  *y = t;
}

void bubbleSort(char arr[], int n)
{
  int i, j;
  for(i = 0; i < n - 1; i++)
    for(j = 0; j < n - i - 1; j++)
      if(strcmp(&arr[j], &arr[j+1]) > 0)
        swap(&arr[j][0], &arr[j+1][0]);
}

void BWT(char *text)
{
  int n = strlen(text);
  char *suffixArray[n];

  //create suffix array
  for(int i = 0; i < n; i++) {
    suffixArray[i] = (char *)malloc(n+1);
    strcpy(suffixArray[i], text + i);
    strcat(suffixArray[i], text);
  }

  // sort suffix array
  bubbleSort(suffixArray, n);

  // extract last character of each suffix 
  printf("transformed text: ");
  for(int i = 0; i < n; i++) {
    printf("%c", suffixArray[i][n-1]);
  }
  printf("\n");

  // free memory
  for (int i = 0; i < n; i++) {
    free(suffixArray[i[);
  }
}
int main()
{
  char text[] = "banana";
  BWT(text);
  return 0;
}

File: javascript/bwt.js
// lossless data compression algorithm
// rearranges input data to improve compression ratios, especially for highly repetitive text

function BWT(text)
{
  const suffixArray = [];
  const n = text.length;

  // create suffix array
  for(let i = 0; i < n; i++)
    {
      suffixArray.push(text.substring(i));
    }
  // sort suffix array lexicographically 
  suffixArray.sort();
  // extract last character of each sorted suffix
  let BWT = "";
  for(let i = 0; i < n; i++)
    {
      BWT += suffixArray[i].charAt(n-1);
    }
  return BWT;
}

// example usage
const text = "banana";
const BWT_text = BWT(text);
console.log("BWT: ", BWT_text);
